### PR TITLE
Support Thread.uncaughtExceptionHandler with OpenJDK

### PR DESCRIPTION
**Pull request by chrisr3**
I have noticed that Avian doesn't support `Thread.uncaughtExceptionHandler` properly when compiled with OpenJDK. ~~This PR _almost_ does what I need except that I have noticed that the test cases are failing because `Thread.join()` isn't waiting for the `uncaughtExceptionHandler` to be executed.~~

~~The test case passes when Avian is _not_ compiled against OpenJDK, e.g.~~

> $ make clean test

~~My test case also passes when executed by both Oracle's JVM and OpenJDK itself, and all JVMs (including Avian) execute `uncaughtExceptionHandler` in the child thread. It therefore seems reasonable for `Thread.join()` to wait until the handler has executed too.~~

I have refactored how this works so that `uncaughtExceptionHandler` is now invoked _before_ `ActiveFlag` is cleared. This allows `Thread.join()` to wait correctly.

### CODE REVIEW

File: src/classpath-openjdk.cpp [hunk 1]
```diff
@@ -975,6 +978,22 @@ class EmbeddedFile {
...
+                                     roots(t)->bootLoader(),
+                                     "java/lang/Thread",
+                                     "dispatchUncaughtException",
+                                     "(Ljava/lang/Throwable;)V");
+  if (dispatch != NULL) {
+    t->m->processor->invoke(t, dispatch, t->javaThread, e);
+
+    // We ignore any exceptions from the uncaught
+    // exception handler itself.
+    t->exception = NULL;
```
Comment (dicej) [inline src/classpath-openjdk.cpp]: I don't think this is enough, since the invoke call above may unwind the stack past it.  Have you tested an exception handler that throws an exception?  I recommend adding such a test to ThreadExceptions.java.  We might need to add a THREAD_RESOURCE0 here to catch the exception and clear it.

Comment (dicej) [reply src/classpath-openjdk.cpp]: e.g. something like this:

```c++
THREAD_RESOURCE0(t, {
  t->exception = NULL;
});

t->m->processor->invoke(t, dispatch, t->javaThread, e);
```

**Comment by chrisr3** [comment]
Not sure what this build error is about, because the failing configuration doesn't have anything to do with `classpath-openjdk.cpp`.

File: src/classpath-openjdk.cpp (hunk 1 above)
Comment (chrisr3) [reply src/classpath-openjdk.cpp]: I see your point: when the `uncaughtExceptionHandler` throws an exception, the `t-checkpoint->unwind()` statement aborts the release of the previous `THREAD_RESOURCE0`. It looks like this might need a custom `Checkpoint` object. Or wrap the method inside a Java `try-catch` block.

**Comment by dicej** [comment]
I'm going to assume the Travis build will succeed when this is merged into master.  If not, we'll fix it.

**Comment by chrisr3** [comment]
Thanks for merging this, although the "exception code path" is rather hard to follow now. I'm not 100% certain how this call-stack unwinding mechanism works, but it looks we'd need to install a new "ignore exceptions" checkpoint before invoking `dispatchUncaughtException()` before my `uncaughtException()` function would be guaranteed to return in all cases. And unfortunately, such a checkpoint would seem to need to be implemented in assembler across multiple architectures!

Ideally, we'd like my new `THREAD_RESOURCE0` block to execute _and then return_ to allow the original `THREAD_RESOURCE0` block to _continue_ executing from where it left off.

**Comment by dicej** [comment]
Yeah, it could definitely be refined.  I'll take another look when I have a chance.  If you look for ueses `vmRun` in the code, you can see examples of how to prevent a thread from unwinding past a given point.  You can think of vmRun as a way to catch exceptions and prevent the stack from unwinding further.

`THREAD_RESOURCE0`, on the other hand, is intended to be used to temporarily interrupt stack unwinding so some code can be run before the unwind is resumed.  I do think a direct call to `vmRun` might be more appropriate in this case, but I haven't had time yet to think it through.
